### PR TITLE
Implement calldata arrays for yul IR code generation.

### LINKED ISSUE #8205: [SolYul] Index access and .length for calldata arrays

### PR DESCRIPTION
**Pull request by ekpyron**
Closes https://github.com/ethereum/solidity/issues/8205
Depends on https://github.com/ethereum/solidity/pull/8257

### CODE REVIEW

**Comment by ekpyron** [comment]
I would have preferred to enable much more of the existing tests, but we're still lacking a lot of features for those unfortunately.

File: libsolidity/codegen/ir/IRGenerationContext.cpp [hunk 1]
```diff
@@ -104,6 +148,14 @@ string IRGenerationContext::variable(Expression const& _expression)
 	string var = "expr_" + to_string(_expression.id());
 	if (size == 1)
 		return var;
+	else if (
+		auto const* arrayType = dynamic_cast<ReferenceType const*>(_expression.annotation().type);
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerationContext.cpp]: We should find a generic way to solve this. I hope you agree that names are much better than numbers.

File: libsolidity/codegen/ir/IRGenerationContext.cpp [hunk 2]
```diff
@@ -104,6 +148,14 @@ string IRGenerationContext::variable(Expression const& _expression)
...
+		arrayType && arrayType->location() == DataLocation::CallData && arrayType->isDynamicallySized()
+	)
+	{
+		solAssert(size == 2, "");
+		return var + "_offset, " + var + "_length";
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerationContext.cpp]:
```suggestion
		return variablePart(_expression, "offset") + ", " + variablePart(_expression, "length"):
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -978,8 +980,30 @@ void IRGeneratorForStatements::endVisit(IndexAccess const& _indexAccess)
...
+					"(" <<
+					m_context.variable(_indexAccess.baseExpression()) <<
+					", " <<
+					expressionAsType(*_indexAccess.indexExpression(), *TypeProvider::uint256()) <<
+					")\n";
+				if (arrayType.isByteArray())
+					defineExpression(_indexAccess) << m_utils.conversionFunction(*arrayType.baseType(), *arrayType.baseType()) << "(calldataload(" << vars.front() << "))\n";
+				else if (arrayType.baseType()->isValueType())
+					defineExpression(_indexAccess) << m_utils.readFromCalldata(*arrayType.baseType()) << "(" << vars.front() << ")\n";
+				else if (arrayType.baseType()->sizeOnStack() > 1)
```
Comment (ekpyron) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: @chriseth these here are the ones I really don't like - this is the workarounds for not having tuple assignments... (if we had them, we could just always use ``defineExpression(_indexAccess) << joinHumanReadable(vars) << "\n";)`` in all cases like this)

File: libsolidity/codegen/ir/IRGenerationContext.cpp (hunk 1 above)
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: The most generic solution I guess would be to have the types themselves return not only their ``sizeOnStack()`` but a list of preferred names for their stack slots...

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 4]
```diff
@@ -908,6 +908,58 @@ string YulUtilFunctions::memoryArrayIndexAccessFunction(ArrayType const& _type)
...
+		("dynamicallySized", _type.isDynamicallySized())
+		("dynamicallyEncodedBase", _type.baseType()->isDynamicallyEncoded())
+		("dynamicallySizedBase", _type.baseType()->isDynamicallySized())
+		("arrayLen",  toCompactHexWithPrefix(_type.length()))
+		("accessCalldataTail", _type.baseType()->isDynamicallyEncoded() ? accessCalldataTailFunction(*_type.baseType()): "")
+		.render();
+	});
+}
+
+string YulUtilFunctions::accessCalldataTailFunction(Type const& _type)
```
Comment (ekpyron) [inline libsolidity/codegen/YulUtilFunctions.cpp:936]: I'm also wondering whether we should just have ``CompilerUtils::accessCalldataTail`` call this one (it does exactly the same)... but currently we don't have the infrastructure for that set up...

File: libsolidity/codegen/ir/IRGenerationContext.cpp (hunk 1 above)
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: Should I go for that or do you have something better in mind?

Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: If it is a small change, we can do it. Then please use a default list of `_1, _2` for all types except for those we already defined names for.

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: Ok, I'll write up something like that in a separate commit.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: `defineExpression(_indexAccess, {"offset", "length"}, vars)`

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 5]
```diff
@@ -978,8 +980,30 @@ void IRGeneratorForStatements::endVisit(IndexAccess const& _indexAccess)
...
+					joinHumanReadable(vars) <<
+					" := " <<
+					m_utils.calldataArrayIndexAccessFunction(arrayType) <<
+					"(" <<
+					m_context.variable(_indexAccess.baseExpression()) <<
+					", " <<
+					expressionAsType(*_indexAccess.indexExpression(), *TypeProvider::uint256()) <<
+					")\n";
+				if (arrayType.isByteArray())
+					defineExpression(_indexAccess) << m_utils.conversionFunction(*arrayType.baseType(), *arrayType.baseType()) << "(calldataload(" << vars.front() << "))\n";
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Does it perform cleanup?

File: libsolidity/codegen/YulUtilFunctions.h [hunk 6]
```diff
@@ -168,9 +168,14 @@ class YulUtilFunctions
 
 	/// @returns the name of a function that returns the calldata address for the
 	/// given array base ref and index.
-	/// signature: (baseRef, index) -> address
+	/// signature: (baseRef, index) -> offset[, length]
 	std::string calldataArrayIndexAccessFunction(ArrayType const& _type);
```
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.h:176]: This was declared but never defined??

File: libsolidity/codegen/ir/IRGenerationContext.cpp [hunk 7]
```diff
@@ -50,6 +50,50 @@ string IRGenerationContext::localVariableName(VariableDeclaration const& _varDec
 	return m_localVariables[&_varDecl];
 }
 
+vector<string> IRGenerationContext::addCalldataVariable(VariableDeclaration const& _varDecl)
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGenerationContext.cpp]: Shouldn't this just be a special case of `addLocalVariable`? Local variables of storage pointer type are also not registered via `addStateVariable`.

Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: And then the variable is "called" `vloc_<name>_<id>`, but its "components" are `vloc_<name>_<id>_offset` and `vloca_<name>_<id>_length`, so we use `localVariableName(var, "offset")` and `localVariableName(var, "length")` to access the two components.

Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: So you do not even have to change `m_localVariables`, but instead these only provide the prefixes.

File: libsolidity/codegen/YulUtilFunctions.h (hunk 6 above)
Comment (ekpyron) [reply libsolidity/codegen/YulUtilFunctions.h:176]: Apparently, yes...

File: libsolidity/codegen/ir/IRGenerationContext.cpp (hunk 7 above)
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: Haha, yeah, I started that way. But then I'd need to change local variables to support multiple stack slots and then ``IRLValue`` as well and then all its derived classes. At least as long as we don't have tuple assignments :-). (With tuple assignments the variable name of the local variable could just be a comma-separated list, but still a single ``std::string``...)

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 5 above)
Comment (ekpyron) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: ``m_utils.conversionFunction(*arrayType.baseType(), *arrayType.baseType())`` does nothing but cleanup - at least that was the idea, but I'll double-check.

File: libsolidity/codegen/ir/IRGenerationContext.cpp (hunk 7 above)
Comment (chriseth) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: Can we talk about this in person tomorrow maybe?

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: Yeah, probably makes sense.

Comment (ekpyron) [reply libsolidity/codegen/ir/IRGenerationContext.cpp]: But yeah - once we have stack slot name information associated with the types, something like that might actually work nicely (and for some reason github only now displays your second and third comment to me).

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `af92f267f68fe7f388b97206e1b70c15cac451e9`:
```
Coding style error:
 libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1000: m_utils.calldataArrayIndexAccessFunction(arrayType) <<
 libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1001: "(" <<
 libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1002: IRVariable(_indexAccess.baseExpression()).commaSeparatedList() <<
 libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1003: ", " <<
 libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1004: expressionAsType(*_indexAccess.indexExpression(), *TypeProvider::uint256()) <<
 libsolidity/codegen/ir/IRGeneratorForStatements.cpp:1005: ")\n";

```
Please check that your changes are working as intended.

**Comment by ekpyron** [comment]
Rebased this on top of https://github.com/ethereum/solidity/pull/8257 now - seems to work.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 8]
```diff
@@ -997,8 +1001,29 @@ void IRGeneratorForStatements::endVisit(IndexAccess const& _indexAccess)
...
+				define(var) <<
+					m_utils.calldataArrayIndexAccessFunction(arrayType) <<
+					"(" <<
+					IRVariable(_indexAccess.baseExpression()).commaSeparatedList() <<
+					", " <<
+					expressionAsType(*_indexAccess.indexExpression(), *TypeProvider::uint256()) <<
+					")\n";
+				if (arrayType.isByteArray())
+					define(_indexAccess) <<
+						m_utils.conversionFunction(*arrayType.baseType(), *arrayType.baseType()) <<
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Could it be that such things are missed with the new code?

**Review by chriseth — DISMISSED** [review]

**Review by chriseth — APPROVED** [review]

**Pull request merged**